Install the previous default bundler version as a regular gem when upgrading rubygems

deivid-rodriguez opened the pull request:
# Description:

The problem is that running `gem update --system` currently wipes out the previous default bundler version from the system. That means applications locked to that previous default bundler version will fail because they will no longer find the bundler version that they need.

`gem update --system` should install the previous bundler version as a regular gem, instead of just overwriting it.

Fixes #2671.
Closes #3044 

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [ ] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

hsbt commented:
@deivid-rodriguez Can you show the details of your expected behavior at https://github.com/rubygems/rubygems/issues/2671#issuecomment-568028825?

I'm still not sure what happened the internal of ruby libdir

deivid-rodriguez commented:
These are simple steps to reproduce the problem:

```
$ gem --version
3.0.6

$ gem list bundler

*** LOCAL GEMS ***

bundler (default: 1.17.3)

$ touch Gemfile
$ bundle
The Gemfile specifies no dependencies
Resolving dependencies...
Bundle complete! 0 Gemfile dependencies, 1 gem now installed.
Bundled gems are installed into `./.bundle`

$ gem update --system
Updating rubygems-update
Successfully installed rubygems-update-3.1.2
Installing RubyGems 3.1.2
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.2
  File: bundler-2.1.2.gem
Bundler 2.1.2 installed
RubyGems 3.1.2 installed
Regenerating binstubs

=== 3.1.2 / 2019-12-20

Minor enhancements:

* Restore non prompting `gem update --system` behavior. Pull request #3040
  by David Rodríguez.
* Show only release notes for new code installed. Pull request #3041 by
  David Rodríguez.
* Inform about installed `bundle` executable after `gem update --system`.
  Pull request #3042 by David Rodríguez.
* Use Bundler 2.1.2. Pull request #3043 by SHIBATA Hiroshi.

Bug fixes:

* Require `uri` in source.rb. Pull request #3034 by mihaibuzgau.
* Fix `gem update --system --force`. Pull request #3035 by David
  Rodríguez.
* Move `require uri` to source_list. Pull request #3038 by mihaibuzgau.

=== 3.1.1 / 2019-12-16

Bug fixes:

* Vendor Bundler 2.1.0 again. The version of Bundler with
  RubyGems 3.1.0 was Bundler 2.1.0.pre.3. Pull request #3029 by
  SHIBATA Hiroshi.

=== 3.1.0 / 2019-12-16

Major enhancements:

* Vendor bundler 2.1. Pull request #3028 by David Rodríguez.

Minor enhancements:

* Check for rubygems.org typo squatting sources. Pull request #2999 by
  Luis Sagastume.
* Refactor remote fetcher. Pull request #3017 by David Rodríguez.
* Lazily load `open3`. Pull request #3001 by David Rodríguez.
* Remove `delegate` dependency. Pull request #3002 by David Rodríguez.
* Lazily load `uri`. Pull request #3005 by David Rodríguez.
* Lazily load `rubygems/gem_runner` during tests. Pull request #3009 by
  David Rodríguez.
* Use bundler to manage development dependencies. Pull request #3012 by
  David Rodríguez.

Bug fixes:

* Remove unnecessary executable flags. Pull request #2982 by David
  Rodríguez.
* Remove configuration that contained a typo. Pull request #2989 by David
  Rodríguez.

Compatibility changes:

* Remove 1.8.7 leftovers. Pull request #2972 by David Rodríguez.
* Deprecate `gem generate_index --modern` and `gem generate_index
  --no-modern`. Pull request #2992 by David Rodríguez.

=== 3.1.0.pre3 / 2019-11-11

Minor enhancements:

* Fix gem pristine not accounting for user installed gems. Pull request
  #2914 by Luis Sagastume.
* Refactor keyword argument test for Ruby 2.7. Pull request #2947 by
  SHIBATA Hiroshi.
* Fix errors at frozen Gem::Version. Pull request #2949 by Nobuyoshi
  Nakada.
* Remove taint usage on Ruby 2.7+. Pull request #2951 by Jeremy Evans.
* Check Manifest.txt is up to date. Pull request #2953 by David Rodríguez.
* Clarify symlink conditionals in tests. Pull request #2962 by David
  Rodríguez.
* Update command line parsing to work under ps. Pull request #2966 by
  David Rodríguez.
* Properly test `Gem::Specifications.stub_for`. Pull request #2970 by
  David Rodríguez.
* Fix Gem::LOADED_SPECS_MUTEX handling for recursive locking. Pull request
  #2985 by MSP-Greg.

=== 3.1.0.pre2 / 2019-10-15

Minor enhancements:

* Optimize Gem::Package::TarReader#each. Pull request #2941 by Jean byroot
  Boussier.
* Time comparison around date boundary. Pull request #2944 by Nobuyoshi
  Nakada.

=== 3.1.0.pre1 / 2019-10-08

Major enhancements:

* Try to use bundler-2.1.0.pre.2. Pull request #2923 by SHIBATA Hiroshi.
* [Require] Ensure -I beats a default gem. Pull request #1868 by Samuel
  Giddins.
* [Specification] Prefer user-installed gems to default gems. Pull request
  #2112 by Samuel Giddins.
* Multifactor authentication for yank command. Pull request #2514 by Qiu
  Chaofan.
* Autoswitch to exact bundler version if present. Pull request #2583 by
  David Rodríguez.

Minor enhancements:

* Fix Gem::Requirement equality comparison when ~> operator is used. Pull
  request #2554 by Grey Baker.
* Don't use a proxy if https_proxy env var is empty. Pull request #2567 by
  Luis Sagastume.
* Fix typo in specs warning. Pull request #2585 by Rui.
* Bin/gem: remove initial empty line. Pull request #2602 by Kenyon Ralph.
* Avoid rdoc hook when it's failed to load rdoc library. Pull request
  #2604 by SHIBATA Hiroshi.
* Refactor get_proxy_from_env logic. Pull request #2611 by Luis Sagastume.
* Allow to easily bisect flaky failures. Pull request #2626 by David
  Rodríguez.
* Fix `--ignore-dependencies` flag not installing platform specific gems.
  Pull request #2631 by David Rodríguez.
* Make `gem install --explain` list platforms. Pull request #2634 by David
  Rodríguez.
* Make `gem update --explain` list platforms. Pull request #2635 by David
  Rodríguez.
* Refactoring install and update explanations. Pull request #2643 by David
  Rodríguez.
* Restore transitiveness of version comparison. Pull request #2651 by
  David Rodríguez.
* Undo requirement sorting. Pull request #2652 by David Rodríguez.
* Update dummy version of Bundler for #2581. Pull request #2584 by SHIBATA
  Hiroshi.
* Ignore to handle the different platform. Pull request #2672 by SHIBATA
  Hiroshi.
* Make Gem::Specification.default_stubs to public methods. Pull request
  #2675 by SHIBATA Hiroshi.
* Sort files and test_files in specifications. Pull request #2524 by
  Christopher Baines.
* Fix comment of Gem::Specification#required_ruby_version=. Pull request
  #2732 by Alex Junger.
* Config_file.rb - update path separator in ENV['GEMRC'] logic. Pull
  request #2735 by MSP-Greg.
* Fix `ruby setup.rb` warnings. Pull request #2737 by David Rodríguez.
* Don't use regex delimiters when searching for a dependency. Pull request
  #2738 by Luis Sagastume.
* Refactor query command. Pull request #2739 by Luis Sagastume.
* Don't remove default spec files from mapping after require. Pull request
  #2741 by David Rodríguez.
* Cleanup base test case. Pull request #2742 by David Rodríguez.
* Simplify Specification#gems_dir. Pull request #2745 by David Rodríguez.
* Fix test warning. Pull request #2746 by David Rodríguez.
* Extract an `add_to_load_path` method. Pull request #2749 by David
  Rodríguez.
* Fix setup command if format_executable is true by default. Pull request
  #2766 by Jeremy Evans.
* Update the certificate files to make the test pass on Debian 10. Pull
  request #2777 by Yusuke Endoh.
* Write to the correct config file(.gemrc). Pull request #2779 by Luis
  Sagastume.
* Fix for large values in UID/GID fields in tar archives. Pull request
  #2780 by Alexey Shein.
* Lazy require stringio. Pull request #2781 by Luis Sagastume.
* Make Gem::Specification#ruby_code handle OpenSSL::PKey::RSA objects.
  Pull request #2782 by Luis Sagastume.
* Fix setup command test for bundler with program_suffix. Pull request
  #2783 by Sorah Fukumori.
* Make sure `rake package` works. Pull request #2787 by David Rodríguez.
* Synchronize access to the Gem::Specification::LOAD_CACHE Hash. Pull
  request #2789 by Benoit Daloze.
* Task to install rubygems to local system. Pull request #2795 by David
  Rodríguez.
* Add an attr_reader to Gem::Installer for the package instance variable.
  Pull request #2796 by Daniel Berger.
* Switch CI script to bash. Pull request #2799 by David Rodríguez.
* Move gemcutter utilities code to Gem::Command. Pull request #2803 by
  Luis Sagastume.
* Add raw spec method to gem package. Pull request #2806 by Luis
  Sagastume.
* Improve `rake package` test error message. Pull request #2815 by David
  Rodríguez.
* Resolve `@@project_dir` from test file paths. Pull request #2843 by
  Nobuyoshi Nakada.
* Remove dead code in Gem::Validator. Pull request #2537 by Ellen Marie
  Dash.
* The date might have advanced since TODAY has been set. Pull request
  #2938 by Nobuyoshi Nakada.
* Remove old ci configurations. Pull request #2917 by SHIBATA Hiroshi.
* Add Gem::Dependency identity. Pull request #2936 by Luis Sagastume.
* Filter dependency type and name strictly. Pull request #2930 by SHIBATA
  Hiroshi.
* Always pass an encoding option to Zlib::GzipReader.wrap. Pull request
  #2933 by Nobuyoshi Nakada.
* Introduce default prerelease requirement. Pull request #2925 by David
  Rodríguez.
* Detect libc version, closes #2918. Pull request #2922 by fauno.
* Use IAM role to extract security-credentials for EC2 instance. Pull
  request #2894 by Alexander Pakulov.
* Improve `gem uninstall --all`. Pull request #2893 by David Rodríguez.
* Use `RbConfig::CONFIG['rubylibprefix']`. Pull request #2889 by Nobuyoshi
  Nakada.
* Build the first gemspec we found if no arguments are passed to gem
  build. Pull request #2887 by Luis Sagastume.
* $LOAD_PATH elements should be real paths. Pull request #2885 by
  Nobuyoshi Nakada.
* Use the standard RUBY_ENGINE_VERSION instead of JRUBY_VERSION. Pull
  request #2864 by Benoit Daloze.
* Cleanup after testing `rake package`. Pull request #2862 by David
  Rodríguez.
* Cherry-pick shushing deprecation warnings from ruby-core. Pull request
  #2861 by David Rodríguez.
* Ext/builder.rb cleanup. Pull request #2849 by Luis Sagastume.
* Fix @ran_rake assignment in builder.rb. Pull request #2850 by Luis
  Sagastume.
* Remove test suite warnings. Pull request #2845 by Luis Sagastume.
* Replace domain parameter with a parameter to suppress suggestions. Pull
  request #2846 by Luis Sagastume.
* Move default specifications dir definition out of BasicSpecification.
  Pull request #2841 by Vít Ondruch.
* There is no usage of @orig_env_* variables in test suite. Pull request
  #2838 by SHIBATA Hiroshi.
* Use File#open instead of Kernel#open in stub_specification.rb. Pull
  request #2834 by Luis Sagastume.
* Simplify #to_ruby code. Pull request #2825 by Nobuyoshi Nakada.
* Add a gem attr to the Gem::Package class. Pull request #2828 by Daniel
  Berger.
* Remove useless TODO comment. Pull request #2818 by Luis Sagastume.

Bug fixes:

* Fix typos in History.txt. Pull request #2565 by Igor Zubkov.
* Remove unused empty sources array. Pull request #2598 by Aaron
  Patterson.
* Fix windows specific executables generated by `gem install`. Pull
  request #2628 by David Rodríguez.
* Gem::Specification#to_ruby needs OpenSSL. Pull request #2937 by
  Nobuyoshi Nakada.
* Set SOURCE_DATE_EPOCH env var if not provided. Pull request #2882 by
  Ellen Marie Dash.
* Installer.rb - fix #windows_stub_script. Pull request #2876 by MSP-Greg.
* Fixed deprecation message. Pull request #2867 by Nobuyoshi Nakada.
* Fix requiring default gems to consider prereleases. Pull request #2728
  by David Rodríguez.
* Forbid `find_spec_for_exe` without an `exec_name`. Pull request #2706 by
  David Rodríguez.
* Do not prompt for passphrase when key can be loaded without it. Pull
  request #2710 by Luis Sagastume.
* Add missing wrapper. Pull request #2690 by David Rodríguez.
* Remove long ago deprecated methods. Pull request #2704 by David
  Rodríguez.
* Renamed duplicate test. Pull request #2678 by Nobuyoshi Nakada.
* File.exists? is deprecated. Pull request #2855 by SHIBATA Hiroshi.
* Fixed to warn with shadowing outer local variable. Pull request #2856 by
  SHIBATA Hiroshi.
* Fix explain with ignore-dependencies. Pull request #2647 by David
  Rodríguez.
* Fix default gem executable installation when folder is not `bin/`. Pull
  request #2649 by David Rodríguez.
* Fix cryptic error on local and ignore-dependencies combination. Pull
  request #2650 by David Rodríguez.

Compatibility changes:

* Remove commented code from command.rb. Pull request #2620 by Luis
  Sagastume.
* Suppress keywords warning. Pull request #2934 by Nobuyoshi Nakada.
* Suppress Ruby 2.7's real kwargs warning. Pull request #2912 by Koichi
  ITO.
* Fix Kernel#warn override. Pull request #2911 by Jeremy Evans.
* Make deprecate Gem::RubyGemsVersion and Gem::ConfigMap. Pull request
  #2857 by SHIBATA Hiroshi.
* Deprecate Gem::RemoteFetcher#fetch_size. Pull request #2833 by Luis
  Sagastume.
* Remove conflict.rb code that was supposed to be removed in Rubygems 3.
  Pull request #2802 by Luis Sagastume.
* Explicitly deprecate `rubyforge_project`. Pull request #2798 by David
  Rodríguez.
* Compatibility cleanups. Pull request #2754 by David Rodríguez.
* Remove `others_possible` activation request param. Pull request #2747 by
  David Rodríguez.
* Remove dependency installer deprecated code. Pull request #2740 by Luis
  Sagastume.
* Deprecate unused Gem::Installer#unpack method. Pull request #2715 by Vít
  Ondruch.
* Removed guard condition with USE_BUNDLER_FOR_GEMDEPS. Pull request #2716
  by SHIBATA Hiroshi.
* Skip deprecation warning during specs. Pull request #2718 by David
  Rodríguez.
* Remove QuickLoader reference. Pull request #2719 by David Rodríguez.
* Removed circular require. Pull request #2679 by Nobuyoshi Nakada.
* Removed needless environmental variable for Travis CI. Pull request
  #2685 by SHIBATA Hiroshi.
* Deprecate a few unused methods. Pull request #2674 by David Rodríguez.
* Removing yaml require. Pull request #2538 by Luciano Sousa.
* Add deprecation warnings for cli options. Pull request #2607 by Luis
  Sagastume.

Style changes:

* Enable `Layout/SpaceInsideParens` rubocop cop. Pull request #2630 by
  David Rodríguez.
* Remove some extra empty lines from the repo. Pull request #2669 by David
  Rodríguez.
* Enable Style/EmptyLinesAroundClassBody rubocop cop. Pull request #2636
  by David Rodríguez.
* Enable Style/BlockDelimiters rubocop cop. Pull request #2640 by David
  Rodríguez.
* Enable Layout/SpaceAroundOperators rubocop cop. Pull request #2642 by
  David Rodríguez.
* Rubocop 0.71. Pull request #2785 by David Rodríguez.


------------------------------------------------------------------------------

RubyGems installed the following executables:
	/home/deivid/.rbenv/versions/2.6.5/bin/gem
	/home/deivid/.rbenv/versions/2.6.5/bin/bundle

RubyGems system software updated

$ bundle
Traceback (most recent call last):
	4: from /home/deivid/.rbenv/versions/2.6.5/bin/ruby_executable_hooks:24:in `<main>'
	3: from /home/deivid/.rbenv/versions/2.6.5/bin/ruby_executable_hooks:24:in `eval'
	2: from /home/deivid/.rbenv/versions/2.6.5/bin/bundle:23:in `<main>'
	1: from /home/deivid/.rbenv/versions/2.6.5/lib/ruby/site_ruby/2.6.0/rubygems.rb:294:in `activate_bin_path'
/home/deivid/.rbenv/versions/2.6.5/lib/ruby/site_ruby/2.6.0/rubygems.rb:275:in `find_spec_for_exe': Could not find 'bundler' (1.17.3) required by your /home/deivid/Code/playground/Gemfile.lock. (Gem::GemNotFoundException)
To update to the latest version installed on your system, run `bundle update --bundler`.
To install the missing version, run `gem install bundler:1.17.3`
```

Updating rubygems removes the old default version of bundler, and people's applications can be depending on it. `gem update --system` should never be destructive in this way.

hsbt commented:
Got it.

> Updating rubygems removes the old default version of bundler,

It's only the gemspec because the default gems shipped ruby installer was located under the stdlib directory. 

If your patch changes with the following behavior, I'm okay to merge this.

```
root@e4bcc426e52d:/# gem list | grep bundler
bundler (default: 2.1.2, default: 1.17.3)
```

If we promote the version `1.17.3` as the nomal gem like `bundler (default: 2.1.2, 1.17.3)`. The users can remove `1.17.3` gemspec with `gem uninstall`. So, the users never restore the version of bundler provided by the ruby installer without reinstallation.

deivid-rodriguez commented:
> It's only the gemspec because the default gems shipped ruby installer was located under the stdlib directory.

It overwrites the `bundler` installation under the stdlib directory, so it effectively removes the old one.

> If your patch changes with the following behavior, I'm okay to merge this.

> ```
> root@e4bcc426e52d:/# gem list | grep bundler
> bundler (default: 2.1.2, default: 1.17.3)
> ```

I think having multiple defaults for the same gem is completely unnecessary, the point of default gems is that the basic functionality is always available, you don't need two copies. Plus, there can only be a single copy of each gem under the stdlib directory, so using such a trick to prevent uninstallation feels like a bad hack, and pretty confusing.  

> If we promote the version 1.17.3 as the nomal gem like bundler (default: 2.1.2, 1.17.3). The users can remove 1.17.3 gemspec with gem uninstall. So, the users never restore the version of bundler provided by the ruby installer without reinstallation.

If user runs `gem update --system`, she wouldn't normally want to uninstall `bundler 1.17.3`, she just wants to upgrade rubygems. If user runs `gem uninstall bundler:1.17.3`, then she wants to uninstall `bundler 1.17.3` and that's fine. I think it's a reasonable behaviour that `gem update --system` installs a newer version of `bundler` as a default, and demotes the old copy to a regular gem that can be removed.

hsbt commented:
> It overwrites the `bundler` installation under the stdlib directory, so it effectively removes the old one.

Is it the current behavior?  I couldn't confirm it with 2.6.5 and `gem update --system`

```
$ /bin/cat lib/ruby/2.6.0/bundler/version.rb
# frozen_string_literal: false

# Ruby 1.9.3 and old RubyGems don't play nice with frozen version strings
# rubocop:disable MutableConstant

module Bundler
  # We're doing this because we might write tests that deal
  # with other versions of bundler and we are unsure how to
  # handle this better.
  VERSION = "1.17.2" unless defined?(::Bundler::VERSION)
(snip)
```

Can you provide your suggested behavior, directory structure and files? I'm the owner of the specification of default gems in ruby core. I need to understand it.

deivid-rodriguez commented:
> Is it the current behavior? I couldn't confirm it with 2.6.5 and gem update --system

Yes, look

```
$ ruby -v
ruby 2.6.5p114 (2019-10-01 revision 67812) [x86_64-linux]

$ gem --version
3.0.6

$ gem list bundler

*** LOCAL GEMS ***

bundler (default: 1.17.3)

$ cat ~/.rbenv/versions/2.6.5/lib/ruby/site_ruby/2.6.0/bundler/version.rb 
# frozen_string_literal: false

# Ruby 1.9.3 and old RubyGems don't play nice with frozen version strings
# rubocop:disable MutableConstant

module Bundler
  # We're doing this because we might write tests that deal
  # with other versions of bundler and we are unsure how to
  # handle this better.
  VERSION = "1.17.3" unless defined?(::Bundler::VERSION)

  def self.overwrite_loaded_gem_version
    begin
      require "rubygems"
    rescue LoadError
      return
    end
    return unless bundler_spec = Gem.loaded_specs["bundler"]
    return if bundler_spec.version == VERSION
    bundler_spec.version = Bundler::VERSION
  end
  private_class_method :overwrite_loaded_gem_version
  overwrite_loaded_gem_version

  def self.bundler_major_version
    @bundler_major_version ||= VERSION.split(".").first.to_i
  end
end
```

I'm on ruby 2.6.5, rubygems 3.0, and I'm only using bundler 1.17.3. My apps could be depending on bundler 1.17.3.

Now I upgrade rubygems with `gem update --system`. As a result:

```
$ gem --version
3.1.2

$ gem list bundler

*** LOCAL GEMS ***

bundler (default: 2.1.2)

$ cat ~/.rbenv/versions/2.6.5/lib/ruby/site_ruby/2.6.0/bundler/version.rb 
# frozen_string_literal: false

module Bundler
  VERSION = "2.1.2".freeze

  def self.bundler_major_version
    @bundler_major_version ||= VERSION.split(".").first.to_i
  end
end
```

Bundler 1.17.3 has been overwritten.

hsbt commented:
> $ cat ~/.rbenv/versions/2.6.5/lib/ruby/site_ruby/2.6.0/bundler/version.rb

`site_ruby is not the standard environment. I said the stdlib is `lib/ruby/2.6.0/bundler/version.rb`.

hsbt commented:
I hope to show your expected specification, Can you show the under the 2.5 and 2.6+ cases?

MSP-Greg commented:
@hsbt said
> bundler (default: 2.1.2, default: 1.17.3)

I would prefer this, if 2.1.2 was installed via `gem update --system` and 1.17.3 was the version installed as a default with Ruby.

But, doing so removes the gemspec for 1.17.3, which seems wrong.  I used a Ruby master/trunk build from June.

deivid-rodriguez commented:
> site_ruby is not the standard environment. I said the stdlib is lib/ruby/2.6.0/bundler/version.rb`.

Oh, sorry, right. Rubygems never touches the stdlib directory, that belongs to ruby. It puts its files in either `rubysitedir` or `rubyvendordir`. However, since `rubysitedir` and `rubyvendordir` take precedence in the LOAD_PATH over the stdlib, the `bundler` version in `stdlib` will be shadowed and won't be used again. I don't think that changes any of the rationale though?

> I hope to show your expected specification, Can you show the under the 2.5 and 2.6+ cases?

I'm not sure what you're asking me. I don't think it makes a difference which ruby version we're talking about. Only thing I'm doing is installing the previous default gem as a regular gem before completely wiping it out.

hsbt commented:
The default gem of bundler on Ruby 2.5(or 2.4...) is only provided under the `rubysitedir` by `gem update --system` But Ruby 2.6+ have default gems of bundler located `rubysitedir` and `rubylibdir`.

The default gems of bundler located on `rubylibdir` never removed by `gem update --system`. I'm not sure what happend with your changes. I hope to understand it.

MSP-Greg commented:
I'm not sure were all on the same page...

@deivid-rodriguez

> I don't think it makes a difference which ruby version we're talking about.

It does if the Ruby version has Bundler as a default gem.

> Only thing I'm doing is installing the previous default gem as a regular gem before shadowing it.

By 'previous default gem', I assume you're referring a version installed by a previous `gem update --system` command?

@hsbt

> The default gems of bundler located on rubylibdir never removed by gem update --system.

`gem update --system` removes the default Bundler gemspec from the default folder.

hsbt commented:
@MSP-Greg Yes, I think the root cause of this issue.

deivid-rodriguez commented:
> gem update --system removes the default Bundler gemspec from the default folder.

Even if it didn't remove it, it wouldn't make a difference, because `rubysitedir` comes before the stdlib folder in the `$LOAD_PATH`, so the version of bundler that comes with ruby, even if the files are still there, can no longer be activated.

MSP-Greg commented:
@deivid-rodriguez

I was just about to type something similar.  There can only be one default gem, because it's resolved from a set of directories, and the first found 'wins'...

deivid-rodriguez commented:
> It does if the Ruby version has Bundler as a default gem.

If no previous default gem for `bundler` is found, my changes do nothing.

> By 'previous default gem', I assume you're referring a version installed by a previous gem update --system command?

Either that, or the version that comes with ruby if no `gem update --system` has been run yet.

> I was just about to type something similar. There can only be one default gem, because it's resolved from a set of directories, and the first found 'wins'...

That's correct :+1:.

deivid-rodriguez commented:
I removed a few changes from the patch to try make it smaller and easier to understand.

MSP-Greg commented on lib/rubygems/remote_fetcher.rb:
```diff
@@ -104,14 +104,14 @@ def initialize(proxy=nil, dns=nil, headers={})
   # larger, more encompassing effort. -erikh
 
-  def download_to_cache(dependency)
+  def download_to_cache(dependency, install_dir = Gem.dir)
```
Personal preference, but maybe change `install_dir` to a kwarg?  They stick out a little more when used...

deivid-rodriguez replied:
I don't want to make it inconsistent with the above `download` method.

deivid-rodriguez commented:
> EDIT: So the only option is to have two copies of the code if Bunder is a default gem in the Ruby installation. Or, write some (possibly) brittle code with symlink...

Yeah. I don't think it's a big deal, although I did open a issue to ruby-core a while ago, because the fact that rubygems and bundler shared its location with the other libs caused some issues: https://bugs.ruby-lang.org/issues/15610. But those issues were fixed so it's not a big deal now IMO.

MSP-Greg commented:
> And same thing goes for rubygems

Hmm.  Bundler is a gem, RubyGems is not.

Dumb question, do we know that older Bundler versions work with more recent versions of RubyGems?

Also, is anyone testing this with older Ruby versions?  Asking because it's relatively easy for me to do (I think)...

MSP-Greg commented:
@hsbt

Did two cases:

<details>
  <summary>Ruby 2.4 with <code>gem update --system -v 3.0.6</code> to <code>gem update --system -v 3.1.3</code></summary>

```
C:\> ruby -v
ruby 2.4.9p362 (2019-10-02 revision 67824) [x64-mingw32]

C:\> gem -v
2.6.14.4

—————————————————————————————————————————————————

C:\> gem update --system 3.0.6
Updating rubygems-update
Fetching: rubygems-update-3.0.6.gem (100%)
Successfully installed rubygems-update-3.0.6
Installing RubyGems 3.0.6
Bundler 1.17.3 installed
RubyGems 3.0.6 installed

RubyGems installed the following executables:
        C:/Greg/Ruby24-x64/bin/gem
        C:/Greg/Ruby24-x64/bin/bundle

RubyGems system software updated

C:\> gem -v
3.0.6

C:\> gem list bundler
*** LOCAL GEMS ***

bundler (default: 1.17.3)

—————————————————————————————————————————————————

C:\> gem install rubygems-update-3.1.3.gem
Successfully installed rubygems-update-3.1.3
1 gem installed

C:\> update_rubygems
Fetching bundler-1.17.3.gem
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.2
  File: bundler-2.1.2.gem
Bundler 2.1.2 installed
RubyGems 3.1.3 installed

C:\> gem -v
3.1.3

C:\> gem list bundler
*** LOCAL GEMS ***

bundler (default: 2.1.2, 1.17.3)
```
</details>

<details>
  <summary>Ruby 2.6 with <code>gem update --system -v 3.1.3</code></summary>

```
PS C:\> ruby -v
ruby 2.6.5p114 (2019-10-01 revision 67812) [x64-mingw32]

PS C:\> gem -v
3.0.3

PS C:\> gem list bundler
*** LOCAL GEMS ***

bundler (default: 1.17.2)

—————————————————————————————————————————————————

PS C:\> gem install rubygems-update-3.1.3.gem
Successfully installed rubygems-update-3.1.3
1 gem installed

PS C:\> update_rubygems
Fetching bundler-1.17.2.gem
  Successfully built RubyGem
  Name: bundler
  Version: 2.1.2
  File: bundler-2.1.2.gem
Bundler 2.1.2 installed
RubyGems 3.1.3 installed
Regenerating binstubs

RubyGems installed the following executables:
        C:/Greg/Ruby26-x64/bin/gem
        C:/Greg/Ruby26-x64/bin/bundle

PS C:\> gem -v
3.1.3

PS C:\> gem list bundler
*** LOCAL GEMS ***

bundler (default: 2.1.2, 1.17.2)
```
</details>

deivid-rodriguez commented:
> Hmm. Bundler is a gem, RubyGems is not.

I'm talking about the copy of them in stdlib getting shadowed by new copies put in `rubysitedir` by rubygems installer. That happens for both whether they are gems or not.

> Did two cases:
> * Ruby 2.4 with gem update --system -v 3.0.6 to gem update --system -v 3.1.3
> * Ruby 2.6 with gem update --system -v 3.1.3

Thanks for confirming that the patch works in those cases. I had tried a few situations in addition to the new unit tests I wrote, but didn't try the 2.4 case for example :+1:.

hsbt commented:
@deivid-rodriguez Is it ok to ready to review?

deivid-rodriguez commented:
Yes, it is!

bronzdoc commented on lib/rubygems/commands/setup_command.rb:
```diff
@@ -392,10 +392,36 @@ def install_default_bundler_gem(bin_dir)
+      unless File.exist?(gem)
+        require 'rubygems/remote_fetcher'
+
+        dep = Gem::Dependency.new spec.name, spec.version
```
Please add parenthesis if is not to much hassle for you, thanks.

bronzdoc commented on lib/rubygems/commands/setup_command.rb:
```diff
@@ -392,10 +392,36 @@ def install_default_bundler_gem(bin_dir)
+        require 'rubygems/remote_fetcher'
+
+        dep = Gem::Dependency.new spec.name, spec.version
+        Gem::RemoteFetcher.fetcher.download_to_cache dep, spec.base_dir
```
same here 😬

hsbt commented:
I considered we can skip the merge this PR because of the users who hope to use rubygems-update in CI environment like Travis CI already added the `--force` flag in their environment.

deivid-rodriguez commented:
This is unrelated to the `--force` flag.

The problem is that `gem update --system` currently removes the previous default `bundler` version from the user's system. In the case of upgrading to the current rubygems, the user will get the old version 1.17.3 removed, and the new one 2.1.2 installed.

Due to the current `BundlerVersionFinder` present in rubygems, that will cause any applications having 1.17.3 in the `BUNDLED WITH` section in the `Gemfile` to not be able to boot any more.

hsbt commented:
@deivid-rodriguez Ah, sorry. I did post the wrong issue for https://github.com/rubygems/rubygems/pull/3045#issuecomment-578454242 . 

https://github.com/rubygems/rubygems/pull/3040 is the correct issue for my comment.

eregon commented:
This seems overly complicated, why not simply do `gem install bundler` at the end of `gem update --system`?
Or even better, don't install bundler at all since the user did not ask for it: https://github.com/rubygems/rubygems/issues/2671#issuecomment-683716488

deivid-rodriguez commented:
Let's keep things as they are now and focus on fixing the issue. I'll rebase this PR.

eregon commented:
Fixing the issue is IMHO either:
* Install bundler with `gem install bundler` during `gem update --system`.
* Do not install bundler with `gem install bundler` during `gem update --system`.

Anything is adding uneeded complexity and hacks on top of hacks.
I don't see any reason to install Bundler in `site_ruby`, that's a hack and it's causing this issue (and many others, I remember some Bundler version accidentally loading some files from another version due to being in `site_ruby`). Let's get rid of the unneeded hack.

deivid-rodriguez commented:
Since bundler 2.3.0 will have the ability to automatically install older bundler versions according to what's in the `BUNDLED WITH` section of the lockfile for each application, now an alternative fix for this issue is #5181. So I'm closing this PR in favor of that one.

eregon commented:
I don't think the problem in https://github.com/rubygems/rubygems/issues/2671#issuecomment-683716488 is addressed by that PR, is it?

deivid-rodriguez commented:
Since the next rubygems update will install bundler 2.3.0 as a default gem, which knows how to "trampoline", I don't think #2671 is a problem anymore as long as we no longer hard error in rubygems if the bundler version does not match the lockfile. At least the realworld end user issue that ticket mentions: "This prevents people from using an older version of bundler on their system if they have a need for it." does not seem an issue to me anymore.

eregon commented:
The fundamental issue is installing Bundler in `site_ruby` shadows the default Bundler, and I think nobody expects `gem update --system` to effectively *remove* the default Bundler version.
I also don't get the point to install bundler into `site_ruby`, it seems to have zero advantages and only problems.

Regarding trampolining, the latest installed bundler is anyway used, whether or not it's in `site_ruby`, no?
Although RubyGems might force an older Bundler if there exact for `BUNDLED WITH` but then there is no need to trampoline anyway.

deivid-rodriguez commented:
Yes, sure, but if the locked major version of bundler does not match the major running version, rubygems currently aborts with an error and there's no chance for bundler to "trampoline back". If we remove this error from rubygems, things will work for example if you have a lockfile with `1.17.3` and only bundler `2.3.0` installed.

Regarding installing bundler as a default gem or not, and where, when rubygems is updated, I think it's fine as is and I strongly disagree with reverting that feature. Maybe I would've hesitated to introduce that in the first place at the stage it was done, but at this point it works just fine and it's very handy, and it has become an established behavior.
